Sources/Provider/CollectionComposer.swift

```diff
@@ -8,145 +8,104 @@
 
 import UIKit
 
-open class CollectionComposer: BaseCollectionProvider {
+open class CollectionComposer: SectionSource, CollectionReloadable {
+
+  public var identifier: String?
 
   public var sections: [AnyCollectionProvider] {
-    get { return dataProvider.data }
-    set { dataProvider.data = newValue }
+    didSet { setNeedsReload() }
   }
 
   public var presenter: CollectionPresenter? {
-    didSet {
-      setNeedsReload()
-    }
+    didSet { setNeedsReload() }
   }
 
-  public var layout: CollectionLayout<AnyCollectionProvider> {
-    didSet {
-      setNeedsReload()
-    }
+  public var layout: CollectionLayout {
+    didSet { setNeedsReload() }
   }
 
-  private var sectionBeginIndex: [Int] = []
-  private var dataProvider: ArrayDataProvider<AnyCollectionProvider>
-
   public init(identifier: String? = nil,
-              layout: CollectionLayout<AnyCollectionProvider> = FlowLayout(),
+              layout: CollectionLayout = FlowLayout(),
               presenter: CollectionPresenter? = nil,
               sections: [AnyCollectionProvider]) {
     self.presenter = presenter
     self.layout = layout
-    self.dataProvider = ArrayDataProvider(data: sections, identifierMapper: {
-      return $1.identifier ?? "\($0)"
-    })
-    super.init(identifier: identifier)
+    self.sections = sections
+    self.identifier = identifier
   }
 
   public convenience init(identifier: String? = nil,
-                          layout: CollectionLayout<AnyCollectionProvider> = FlowLayout(),
+                          layout: CollectionLayout = FlowLayout(),
                           presenter: CollectionPresenter? = nil,
                           _ sections: AnyCollectionProvider...) {
-    self.init(layout: layout, presenter: presenter, sections: sections)
+    self.init(identifier: identifier, layout: layout, presenter: presenter, sections: sections)
   }
 
-  func indexPath(_ index: Int) -> (Int, Int) {
-    let sectionIndex = sectionBeginIndex.binarySearch { $0 <= index } - 1
-    return (sectionIndex, index - sectionBeginIndex[sectionIndex])
+  open var numberOfItems: Int {
+    return sections.count
   }
 
-  open override var numberOfItems: Int {
-    return (sectionBeginIndex.last ?? 0) + (sections.last?.numberOfItems ?? 0)
+  open func section(at: Int) -> AnyCollectionProvider? {
+    return sections[at]
   }
 
-  open override func view(at: Int) -> UIView {
-    let (sectionIndex, item) = indexPath(at)
-    return sections[sectionIndex].view(at: item)
+  open func identifier(at: Int) -> String {
+    return sections[at].identifier ?? "\(at)"
   }
 
-  open override func update(view: UIView, at: Int) {
-    let (sectionIndex, item) = indexPath(at)
-    sections[sectionIndex].update(view: view, at: item)
+  open func layout(collectionSize: CGSize) {
+    layout.layout(context: CollectionComposerLayoutContext(collectionSize: collectionSize,
```

Comment:
Use a property to store this `CollectionComposerLayoutContext`? So that if `collectionSize` is the same we can avoid a layout calculation pass?

Reply:
Since it is a struct, it doesn't incur a lot of performance cost.